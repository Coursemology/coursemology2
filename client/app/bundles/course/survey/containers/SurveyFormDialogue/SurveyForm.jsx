import PropTypes from 'prop-types';
import {
  defineMessages,
  injectIntl,
  intlShape,
  FormattedMessage,
} from 'react-intl';
import { reduxForm, Field, Form } from 'redux-form';
import ReactTooltip from 'react-tooltip';
import renderTextField from 'lib/components/redux-form/TextField';
import DateTimePicker from 'lib/components/redux-form/DateTimePicker';
import renderToggleField from 'lib/components/redux-form/Toggle';
import formTranslations from 'lib/translations/form';
import translations from 'course/survey/translations';
import { formNames } from 'course/survey/constants';

const styles = {
  columns: {
    display: 'flex',
  },
  oneColumn: {
    flex: 1,
  },
  toggle: {
    marginTop: 16,
  },
  hint: {
    fontSize: 14,
    marginBottom: 12,
  },
};

const surveyFormTranslations = defineMessages({
  startEndValidationError: {
    id: 'course.surveys.SurveyForm.startEndValidationError',
    defaultMessage: "Must be after 'Opens At'",
  },
  allowResponseAfterEndHint: {
    id: 'course.surveys.SurveyForm.allowResponseAfterEndHint',
    defaultMessage:
      'Allow students to submit responses after the survey has expired. \
      If this is enabled, students who submit before the deadline will get both the base and bonus \
      points, whereas students who submit after the deadline will only be awarded the base points.',
  },
  allowModifyAfterSubmitHint: {
    id: 'course.surveys.SurveyForm.allowModifyAfterSubmitHint',
    defaultMessage:
      'Allow students to modify their responses after they have submitted it. If \
      this is disabled, you will have to manually unsubmit their responses to allow them to \
      edit it.',
  },
  anonymousHint: {
    id: 'course.surveys.SurveyForm.anonymousHint',
    defaultMessage:
      'If you make the survey anonymous, you will be able to see aggregate survey \
      results but not individual responses. You may not toggle this setting once there is one \
      or more student submissions.',
  },
  hasStudentResponse: {
    id: 'course.surveys.SurveyForm.hasStudentResponse',
    defaultMessage:
      'At least one student has responded to this survey. You may not remove anonymity.',
  },
  timeBonusExpTooltip: {
    id: 'course.surveys.SurveyForm.timeBonusExpTooltip',
    defaultMessage:
      'You must allow responses after the survey expires to set bonus points.',
  },
});

const validate = (values) => {
  const errors = {};

  const requiredFields = ['title', 'start_at'];
  requiredFields.forEach((field) => {
    if (!values[field]) {
      errors[field] = formTranslations.required;
    }
  });

  if (values.end_at && new Date(values.start_at) >= new Date(values.end_at)) {
    errors.end_at = surveyFormTranslations.startEndValidationError;
  }

  if (values.allow_response_after_end && !values.end_at) {
    errors.end_at = formTranslations.required;
  }

  return errors;
};

const propTypes = {
  formValues: PropTypes.shape({
    start_at: PropTypes.instanceOf(Date),
    end_at: PropTypes.instanceOf(Date),
    allow_response_after_end: PropTypes.bool,
  }),
  shiftEndDate: PropTypes.func.isRequired,
  handleSubmit: PropTypes.func.isRequired,
  onSubmit: PropTypes.func.isRequired,
  intl: intlShape.isRequired,
  disableAnonymousToggle: PropTypes.bool,
  disabled: PropTypes.bool,
};

const SurveyForm = ({
  handleSubmit,
  intl,
  onSubmit,
  disabled,
  disableAnonymousToggle,
  shiftEndDate,
  formValues,
}) => (
  <Form onSubmit={handleSubmit(onSubmit)}>
    <Field
      fullWidth
      name="title"
      label={intl.formatMessage(translations.title)}
      component={renderTextField}
      {...{ disabled }}
    />
    <br />
    <Field
      fullWidth
      name="description"
      label={intl.formatMessage(translations.description)}
      component={renderTextField}
      multiline
      rows={2}
      {...{ disabled }}
    />
    <div style={styles.columns}>
      <Field
        name="start_at"
        label={intl.formatMessage(translations.opensAt)}
        component={DateTimePicker}
        afterChange={(_, newStartAt) =>
          shiftEndDate(formNames.SURVEY, newStartAt, formValues)
        }
        style={styles.oneColumn}
        {...{ disabled }}
      />
      <Field
        name="end_at"
        label={intl.formatMessage(translations.expiresAt)}
        component={DateTimePicker}
        style={styles.oneColumn}
        {...{ disabled }}
      />
    </div>
    <div style={styles.columns}>
      <div style={styles.oneColumn}>
        <Field
          name="base_exp"
          label={intl.formatMessage(translations.basePoints)}
          component={renderTextField}
          type="number"
          onWheel={(event) => event.currentTarget.blur()}
          {...{ disabled }}
        />
      </div>
      <div
        style={styles.oneColumn}
        data-tip
        data-for="timeBonusExpTooltip"
        data-tip-disable={formValues && formValues.allow_response_after_end}
      >
        <Field
          name="time_bonus_exp"
          label={intl.formatMessage(translations.bonusPoints)}
          component={renderTextField}
          type="number"
          onWheel={(event) => event.currentTarget.blur()}
          disabled={formValues && !formValues.allow_response_after_end}
        />
        <ReactTooltip id="timeBonusExpTooltip">
          <FormattedMessage {...surveyFormTranslations.timeBonusExpTooltip} />
        </ReactTooltip>
      </div>
    </div>
    <Field
      name="allow_response_after_end"
      component={renderToggleField}
      parse={Boolean}
      label={intl.formatMessage(translations.allowResponseAfterEnd)}
      style={styles.toggle}
      disabled={disabled}
    />
    <div style={styles.hint}>
      {intl.formatMessage(surveyFormTranslations.allowResponseAfterEndHint)}
    </div>
    <Field
      name="allow_modify_after_submit"
      component={renderToggleField}
      parse={Boolean}
      label={intl.formatMessage(translations.allowModifyAfterSubmit)}
      style={styles.toggle}
      disabled={disabled}
    />
    <div style={styles.hint}>
      {intl.formatMessage(surveyFormTranslations.allowModifyAfterSubmitHint)}
    </div>
    <Field
      name="anonymous"
      component={renderToggleField}
      parse={Boolean}
      label={intl.formatMessage(translations.anonymous)}
      style={styles.toggle}
      disabled={disableAnonymousToggle || disabled}
    />
    <div style={styles.hint}>
      {disableAnonymousToggle
        ? intl.formatMessage(surveyFormTranslations.hasStudentResponse)
        : intl.formatMessage(surveyFormTranslations.anonymousHint)}
    </div>
  </Form>
);

SurveyForm.propTypes = propTypes;

export default reduxForm({
  form: formNames.SURVEY,
  validate,
})(injectIntl(SurveyForm));
