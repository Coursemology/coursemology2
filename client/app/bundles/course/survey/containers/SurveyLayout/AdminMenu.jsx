/* eslint-disable camelcase */
import { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { injectIntl, defineMessages, intlShape } from 'react-intl';
import { IconButton, Menu, MenuItem } from '@material-ui/core';
import MoreVert from '@material-ui/icons/MoreVert';
import * as surveyActions from 'course/survey/actions/surveys';
import { showDeleteConfirmation } from 'course/survey/actions';
import { formatSurveyFormData } from 'course/survey/utils';
import { surveyShape } from 'course/survey/propTypes';

const translations = defineMessages({
  editSurvey: {
    id: 'course.surveys.SurveyLayout.AdminMenu.editSurvey',
    defaultMessage: 'Edit Survey',
  },
  deleteSurvey: {
    id: 'course.surveys.SurveyLayout.AdminMenu.deleteSurvey',
    defaultMessage: 'Delete Survey',
  },
  updateSuccess: {
    id: 'course.surveys.SurveyLayout.AdminMenu.updateSuccess',
    defaultMessage: 'Survey "{title}" updated.',
  },
  updateFailure: {
    id: 'course.surveys.SurveyLayout.AdminMenu.updateFailure',
    defaultMessage: 'Failed to update survey.',
  },
  deleteSuccess: {
    id: 'course.surveys.SurveyLayout.AdminMenu.deleteSuccess',
    defaultMessage: 'Survey "{title}" deleted.',
  },
  deleteFailure: {
    id: 'course.surveys.SurveyLayout.AdminMenu.deleteFailure',
    defaultMessage: 'Failed to delete survey.',
  },
});

class AdminMenu extends Component {
  constructor(props) {
    super(props);

    this.state = {
      anchorEl: null,
    };
  }

  deleteSurveyHandler = () => {
    const { survey, dispatch, intl, surveyId } = this.props;
    const { deleteSurvey } = surveyActions;
    const successMessage = intl.formatMessage(
      translations.deleteSuccess,
      survey,
    );
    const failureMessage = intl.formatMessage(translations.deleteFailure);
    const handleDelete = () =>
      dispatch(deleteSurvey(surveyId, successMessage, failureMessage));
    return dispatch(showDeleteConfirmation(handleDelete));
  };

  handleClick = (event) => {
    // This prevents ghost click.
    event.preventDefault();

    this.setState({
      anchorEl: event.currentTarget,
    });
  };

  handleClose = () => {
    this.setState({ anchorEl: null });
  };

  showEditSurveyForm = () => {
    const { survey, dispatch, intl } = this.props;
    const { showSurveyForm } = surveyActions;
    const {
      title,
      description,
      base_exp,
      time_bonus_exp,
      start_at,
      end_at,
      hasStudentResponse,
      allow_response_after_end,
      allow_modify_after_submit,
      anonymous,
    } = survey;

    const initialValues = {
      title,
      description,
      base_exp,
      time_bonus_exp,
      allow_response_after_end,
      allow_modify_after_submit,
      anonymous,
    };

    return dispatch(
      showSurveyForm({
        onSubmit: this.updateSurveyHandler,
        formTitle: intl.formatMessage(translations.editSurvey),
        hasStudentResponse,
        initialValues: {
          ...initialValues,
          start_at: new Date(start_at),
          end_at: end_at && new Date(end_at),
        },
      }),
    );
  };

  updateSurveyHandler = (data) => {
    const { dispatch, intl, surveyId } = this.props;
    const { updateSurvey } = surveyActions;

    const payload = formatSurveyFormData(data);
    const successMessage = intl.formatMessage(translations.updateSuccess, data);
    const failureMessage = intl.formatMessage(translations.updateFailure);
    return dispatch(
      updateSurvey(surveyId, payload, successMessage, failureMessage),
    );
  };

  render() {
    const { intl, survey } = this.props;
    if (!survey.canUpdate && !survey.canDelete) {
      return null;
    }

    return (
      <>
        <IconButton onClick={this.handleClick}>
          <MoreVert htmlColor="white" />
        </IconButton>
        <Menu
          id="admin-menu"
          anchorEl={this.state.anchorEl}
          disableAutoFocusItem
          onClick={this.handleClose}
          onClose={this.handleClose}
          open={Boolean(this.state.anchorEl)}
        >
          {survey.canUpdate && (
            <MenuItem onClick={this.showEditSurveyForm}>
              {intl.formatMessage(translations.editSurvey)}
            </MenuItem>
          )}
          {survey.canDelete && (
            <MenuItem onClick={this.deleteSurveyHandler}>
              {intl.formatMessage(translations.deleteSurvey)}
            </MenuItem>
          )}
        </Menu>
      </>
    );
  }
}

AdminMenu.propTypes = {
  survey: surveyShape,
  surveyId: PropTypes.string.isRequired,

  intl: intlShape,
  dispatch: PropTypes.func.isRequired,
};

export default connect()(injectIntl(AdminMenu));
