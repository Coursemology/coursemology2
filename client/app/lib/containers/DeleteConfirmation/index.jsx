import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import ConfirmationDialog from 'lib/components/ConfirmationDialog';
import { resetDeleteConfirmation } from 'lib/actions';

const DeleteConfirmation = ({ dispatch, deleteConfirmation }) => (
  <ConfirmationDialog
    confirmDelete
    {...deleteConfirmation}
    onCancel={() => dispatch(resetDeleteConfirmation())}
  />
);

DeleteConfirmation.propTypes = {
  dispatch: PropTypes.func.isRequired,
  deleteConfirmation: PropTypes.shape({
    open: PropTypes.bool.isRequired,
    onConfirm: PropTypes.func.isRequired,
  }).isRequired,
};

export default connect((state) => ({
  deleteConfirmation: state.deleteConfirmation,
}))(DeleteConfirmation);
