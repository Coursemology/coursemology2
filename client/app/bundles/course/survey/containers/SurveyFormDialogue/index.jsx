import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import * as actionCreators from 'course/survey/actions/surveys';
import FormDialogue from 'lib/components/FormDialogue';
import SurveyForm from './SurveyForm';

function mapStateToProps({ surveyForm, ...state }) {
  return {
    ...surveyForm,
    ...state,
  };
}

const propTypes = {
  dispatch: PropTypes.func.isRequired,
  visible: PropTypes.bool.isRequired,
  hasStudentResponse: PropTypes.bool.isRequired,
  disabled: PropTypes.bool.isRequired,
  formTitle: PropTypes.string,
  onSubmit: PropTypes.func.isRequired,
  initialValues: PropTypes.shape({
    title: PropTypes.string,
    description: PropTypes.string,
    start_at: PropTypes.instanceOf(Date),
    end_at: PropTypes.instanceOf(Date),
    base_exp: PropTypes.number,
    allow_response_after_end: PropTypes.bool,
  }).isRequired,
};

const SurveyFormDialogue = ({
  dispatch,
  visible,
  disabled,
  formTitle,
  hasStudentResponse,
  initialValues,
  onSubmit,
}) => {
  const { hideSurveyForm } = bindActionCreators(actionCreators, dispatch);

  const surveyFormProps = {
    disabled,
    disableAnonymousToggle:
      // eslint-disable-next-line react/prop-types
      initialValues && initialValues.anonymous && hasStudentResponse,
    initialValues,
    onSubmit,
  };

  return (
    <FormDialogue
      title={formTitle}
      open={visible}
      skipConfirmation={false}
      disabled={disabled}
      form="survey-form"
      hideForm={hideSurveyForm}
    >
      <SurveyForm {...surveyFormProps} />
    </FormDialogue>
  );
};

SurveyFormDialogue.propTypes = propTypes;

export default connect(mapStateToProps)(SurveyFormDialogue);
