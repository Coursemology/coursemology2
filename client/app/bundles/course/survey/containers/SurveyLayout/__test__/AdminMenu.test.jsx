import ReactDOM from 'react-dom';
import ReactTestUtils from 'react-dom/test-utils';
import { mount } from 'enzyme';
import CourseAPI from 'api/course';
import DeleteConfirmation from 'lib/containers/DeleteConfirmation';
import storeCreator from 'course/survey/store';
import SurveyFormDialogue from 'course/survey/containers/SurveyFormDialogue';
import AdminMenu from '../AdminMenu';

describe('<AdminMenu />', () => {
  it('does not render button if user cannot edit or update', () => {
    const survey = {
      id: 2,
      title: 'Survey',
      canDelete: false,
      canUpdate: false,
    };
    const adminMenu = mount(
      <AdminMenu survey={survey} surveyId={survey.id.toString()} />,
      buildContextOptions(storeCreator({})),
    );

    expect(adminMenu).toMatchSnapshot();
  });

  it('allows surveys to be deleted', () => {
    const spyDelete = jest.spyOn(CourseAPI.survey.surveys, 'delete');
    const survey = {
      id: 2,
      title: 'Survey To Delete',
      canDelete: true,
    };

    const contextOptions = buildContextOptions(storeCreator({}));
    const deleteConfirmation = mount(<DeleteConfirmation />, contextOptions);
    const adminMenu = mount(
      <AdminMenu survey={survey} surveyId={survey.id.toString()} />,
      contextOptions,
    );

    const iconButton = adminMenu.find('button').first();
    iconButton.simulate('click');

    const deleteButton = adminMenu
      .find('WithStyles(ForwardRef(ButtonBase))')
      .last();
    deleteButton.simulate('click');

    const confirmDeleteButton = deleteConfirmation
      .find('ConfirmationDialog')
      .first()
      .instance().confirmButton;
    ReactTestUtils.Simulate.click(ReactDOM.findDOMNode(confirmDeleteButton));

    expect(spyDelete).toHaveBeenCalledWith(survey.id.toString());
  });

  it('allows surveys to be edited', () => {
    const spyUpdate = jest.spyOn(CourseAPI.survey.surveys, 'update');
    const surveyFormData = {
      title: 'Survey To Edit',
      base_exp: 100,
      time_bonus_exp: 50,
      start_at: '2017-02-27T00:00:00.000+08:00',
      end_at: '2017-03-12T23:59:00.000+08:00',
      allow_response_after_end: true,
      allow_modify_after_submit: true,
      anonymous: true,
    };
    const survey = {
      ...surveyFormData,
      id: 2,
      hasStudentResponse: true,
      canUpdate: true,
    };

    const contextOptions = buildContextOptions(storeCreator({}));
    const surveyFormDialogue = mount(<SurveyFormDialogue />, contextOptions);
    const adminMenu = mount(
      <AdminMenu survey={survey} surveyId={survey.id.toString()} />,
      contextOptions,
    );

    const iconButton = adminMenu.find('button').first();
    iconButton.simulate('click');

    const updateButton = adminMenu
      .find('WithStyles(ForwardRef(ButtonBase))')
      .last();
    updateButton.simulate('click');

    surveyFormDialogue.update();

    const sectionForm = surveyFormDialogue.find('form');
    const description = 'To update description';
    const descriptionInput = sectionForm.find('textarea[name="description"]');
    descriptionInput.simulate('change', { target: { value: description } });

    const submitButton = surveyFormDialogue
      .find('FormDialogue')
      .first()
      .instance().submitButton;
    ReactTestUtils.Simulate.click(ReactDOM.findDOMNode(submitButton));

    const expectedPayload = {
      survey: {
        ...surveyFormData,
        description,
        start_at: new Date(survey.start_at),
        end_at: new Date(survey.end_at),
        bonus_end_at: new Date(survey.end_at),
      },
    };
    expect(spyUpdate).toHaveBeenCalledWith(
      survey.id.toString(),
      expectedPayload,
    );
  });
});
