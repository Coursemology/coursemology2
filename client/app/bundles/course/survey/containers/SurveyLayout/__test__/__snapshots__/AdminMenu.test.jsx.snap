// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AdminMenu /> does not render button if user cannot edit or update 1`] = `
<Connect(injectIntl(AdminMenu))
  survey={
    Object {
      "canDelete": false,
      "canUpdate": false,
      "id": 2,
      "title": "Survey",
    }
  }
  surveyId="2"
>
  <injectIntl(AdminMenu)
    dispatch={[Function]}
    survey={
      Object {
        "canDelete": false,
        "canUpdate": false,
        "id": 2,
        "title": "Survey",
      }
    }
    surveyId="2"
  >
    <AdminMenu
      dispatch={[Function]}
      intl={
        Object {
          "defaultFormats": Object {},
          "defaultLocale": "en",
          "formatDate": [Function],
          "formatDateToParts": [Function],
          "formatDisplayName": [Function],
          "formatList": [Function],
          "formatMessage": [Function],
          "formatNumber": [Function],
          "formatNumberToParts": [Function],
          "formatPlural": [Function],
          "formatRelativeTime": [Function],
          "formatTime": [Function],
          "formatTimeToParts": [Function],
          "formats": Object {},
          "formatters": Object {
            "getDateTimeFormat": [Function],
            "getDisplayNames": [Function],
            "getListFormat": [Function],
            "getMessageFormat": [Function],
            "getNumberFormat": [Function],
            "getPluralRules": [Function],
            "getRelativeTimeFormat": [Function],
          },
          "locale": "en",
          "messages": Object {},
          "onError": [Function],
          "textComponent": Symbol(react.fragment),
          "timeZone": "Asia/Singapore",
          "wrapRichTextChunksInFragment": undefined,
        }
      }
      survey={
        Object {
          "canDelete": false,
          "canUpdate": false,
          "id": 2,
          "title": "Survey",
        }
      }
      surveyId="2"
    />
  </injectIntl(AdminMenu)>
</Connect(injectIntl(AdminMenu))>
`;
